pyteal/ast/abi/type.py

```diff
@@ -1,71 +1,82 @@
 
+class TypeSpec(ABC):
+    """TypeSpec represents a specification for an ABI type.
 
-class Type(ABC):
-    """The abstract base class for all ABI types.
-
-    This class contains both information about an ABI type, and a value that conforms to that type.
-    The value is contained in a unique ScratchVar that only the type has access to. As a result, the
-    value of an ABI type is mutable and can be efficiently referenced multiple times without needing
-    to recompute it.
+    Essentially this is a factory that can produce specific instances of ABI types.
     """
 
-    def __init__(self, valueType: TealType) -> None:
-        """Create a new Type.
-
-        Args:
-            valueType: The TealType (uint64 or bytes) that this ABI type will store in its internal
-                ScratchVar.
-        """
-        super().__init__()
-        self.stored_value = ScratchVar(valueType)
-
     @abstractmethod
-    def has_same_type_as(self, other: "Type") -> bool:
-        """Check if this type is considered equal to the other ABI type, irrespective of their
-        values.
+    def new_instance(self) -> "BaseType":
+        """Create a new instance of the specified type."""
+        pass
 
-        Args:
-            other: The ABI type to compare to.
+    @abstractmethod
+    def is_dynamic(self) -> bool:
+        """Check if this ABI type is dynamic.
 
-        Returns:
-            True if and only if self and other can store the same ABI value.
+        If a type is dynamic, the length of its encoding depends on its value. Otherwise, the type
+        is considered static (not dynamic).
         """
         pass
 
     @abstractmethod
-    def new_instance(self: T) -> T:
-        """Create a new instance of this ABI type.
+    def byte_length_static(self) -> int:
+        """Get the byte length of this ABI type's encoding. Only valid for static types."""
+        pass
 
-        The value of this type will not be applied to the new type.
-        """
+    @abstractmethod
+    def storage_type(self) -> TealType:
+        """Get the TealType that the underlying ScratchVar should hold for this type."""
         pass
 
     @abstractmethod
-    def is_dynamic(self) -> bool:
-        """Check if this ABI type is dynamic.
+    def __eq__(self, other: object) -> bool:
+        """Check if this type is considered equal to another ABI type.
 
-        If a type is dynamic, the length of its encoding depends on its value. Otherwise, the type
-        is considered static (not dynamic).
+        Args:
+            other: The object to compare to. If this is not a TypeSpec, this method will never
+                return true.
+
+        Returns:
+            True if and only if self and other represent the same ABI type.
         """
         pass
 
     @abstractmethod
-    def byte_length_static(self) -> int:
-        """Get the byte length of this ABI type's encoding. Only valid for static types."""
+    def __str__(self) -> str:
```

Comment:
@jasonpaulos Leaving as a sanity check:  While it's convenient today to use `__str__` for ABI method signature encoding, are there longer term downsides to using `__str__` for a protocol specific concern?

Rationale:
* Conventions may well be different here, which is why I pose the question as a sanity check.  In my prior experience (e.g. JSON representation), I found it preferable to define a 1st class representation (e.g. type class) over using the language's `__str__` equivalent.
* Here, I'd nominally lean towards a distinct interface method (e.g. `as_abi_string` or similar).  
* Advantages:
  * Allows `__str__` to be used for arbitrary purposes without breaking ABI encoding.   It's _possible_ someone will prefer a different representation for debugging vs method encoding.
  * Potentially simplifies future refactoring by forcing an explicit method call at all call sites.  I imagine Python will take the `__str__` representation in some use cases without invoking `str()`.
  * Reduces the chance of accidentally forgetting to implement or incorrectly implementing the string representation.  It's subjective, though I think an explicitly named method provides a stronger hint.
    * I see the usage of `pass` prevents a subclass from skipping `__str__` definition.  So, the PR addresses some of the concern.
* The obvious disadvantage is the added verbosity due to explicitly qualifying each conversion.  However, I tend to think the explicit invocation is a desirable correctness quality.

Reply:
Those are valid concerns, but I don't think the downsides of using `__str__` for this are significant. I've tried to address some concerns below.
* In Python there are actually two object-to-string functions, `__str__` and `__repr__`. `__str__` is normally used only by user code that explicitly calls `str(x)`, while `__repr__` is used for language-level object representation in debugging, logging, etc.
* The tradeoff between `__str__` and a decided method is really convenience vs specificity; I don't think either would have technical disadvantages. My preference is to keep it as `__str__`, but I'm open to further discussion.

Reply:
@jasonpaulos Got it - thanks for the context here.  Resolving.

* The difference between `__str__` and `__repr__` helps ease _some_ of my concern.
* I haven't totally shaken the feeling that `__str__` will prove problematic, but I don't feel strongly enough to warrant going deeper.